knative/deployer.go

```diff
@@ -24,52 +25,151 @@ type Deployer struct {
+				}
+				return err
+			}
+
+			fmt.Println("Function deployed on: " + route.Status.URL.String())
+
+		} else {
+			if !d.Verbose {
+				err = fmt.Errorf("deployer failed to get the service: %v.\nStdOut: %s", err, output.(*bytes.Buffer).String())
+			} else {
+				err = fmt.Errorf("deployer failed to get the service: %v", err)
+			}
+			return err
+		}
+	} else {
+		// Update the existing Service
+		err = client.UpdateServiceWithRetry(encodedName, updateBuiltTimeStampEnvVar, 3)
+		if err != nil {
+			if !d.Verbose {
+				err = fmt.Errorf("deployer failed to update the service: %v.\nStdOut: %s", err, output.(*bytes.Buffer).String())
+			} else {
+				err = fmt.Errorf("deployer failed to update the service: %v", err)
+			}
+			return err
+		}
 	}
-	c.SetOut(output)
-	args := []string{
-		"service", "create", encodedName,
-		"--image", f.Image,
-		"--env", "VERBOSE=true",
-		"--label", "bosonFunction=true",
+
+	return nil
+}
+
+func generateNewService(name, image string) *servingv1.Service {
+	containers := []corev1.Container{
+		{
+			Image: image,
+			Env: []corev1.EnvVar{
+				{Name: "VERBOSE", Value: "true"},
+			},
+		},
 	}
-	if d.Namespace != "" {
-		args = append(args, "--namespace", d.Namespace)
+
+	return &v1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name,
+			Labels: map[string]string{
+				"bosonFunction": "true",
+			},
+		},
+		Spec: v1.ServiceSpec{
+			ConfigurationSpec: v1.ConfigurationSpec{
+				Template: v1.RevisionTemplateSpec{
+					Spec: v1.RevisionSpec{
+						PodSpec: corev1.PodSpec{
+							Containers: containers,
+						},
+					},
+				},
+			},
+		},
 	}
-	c.SetArgs(args)
-	err = c.Execute()
-	if err != nil {
-		if !d.Verbose {
-			err = fmt.Errorf("failed to deploy the service: %v.\nStdOut: %s", err, output.(*bytes.Buffer).String())
-		} else {
-			err = fmt.Errorf("failed to deploy the service: %v", err)
+}
+
+func updateBuiltTimeStampEnvVar(service *servingv1.Service) (*servingv1.Service, error) {
+	envs := service.Spec.Template.Spec.Containers[0].Env
+
+	builtEnvVarName := "BUILT"
+
+	builtEnvVar := findEnvVar(builtEnvVarName, envs)
+	if builtEnvVar == nil {
+		envs = append(envs, corev1.EnvVar{Name: "VERBOSE", Value: "true"})
+		builtEnvVar = &envs[len(envs)-1]
+	}
+
+	builtEnvVar.Value = time.Now().Format("20060102T150405")
+
+	sort.SliceStable(envs, func(i, j int) bool {
+		return envs[i].Name <= envs[j].Name
+	})
+	service.Spec.Template.Spec.Containers[0].Env = envs
+
+	return service, nil
+}
+
+func findEnvVar(name string, envs []corev1.EnvVar) *corev1.EnvVar {
+	var result *corev1.EnvVar = nil
+	for i, envVar := range envs {
+		if envVar.Name == name {
+			result = &envs[i]
+			break
 		}
-		return
 	}
-	// TODO: use the KN service client noted above, such that we can return the
-	// final path/route of the final deployed Function.  While it can be assumed
-	// due to being deterministic, new users would be aided by having it echoed.
-	return
+	return result
```

Comment:
@lance @rhuss this will let cluster to generate revision name so this: https://github.com/knative/serving/issues/9544 won't affect us, right?

Reply:
I think so? I'm starting to get confused about the different approaches. Is it correct to use the [commands](https://github.com/boson-project/faas/pull/155/files) or the client as in this case?

Reply:
My impression was, that we should use the client, but I might be wrong.

@rhuss wdyt?

Reply:
I recommend using the client APIs. However *not* the commands but the client API which comes as an interface as here --> https://github.com/knative/client/blob/master/pkg/serving/v1/client.go (this is _not_ the serving API but a facade to it)

Please don't use the commands. They can change anytime in incompatible way, there are no precautions to prevent this.

Also we are planning to expose the client API (not the commands!), the way how to parse options, how to parse sink prefixes, ... and generally anything that can be useful for implementing an plugin into a separate repository (`knative/client-pkg` as working title), so that you don't even have a direct dependency to `knative/client` anymore (which then also prevents you technically to access the cobra kn commands)

Reply:
I tested this and K_SINK is injected, unlike before this change :+1: .